clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -521,64 +530,105 @@ public ConsumerRebalanceListener rebalanceListener() {
             this.nextAllowedRetryTimeMs = null;
         }
 
-        private void reset(OffsetResetStrategy strategy) {
+        private synchronized void reset(OffsetResetStrategy strategy) {
             this.resetStrategy = strategy;
             this.position = null;
             this.nextAllowedRetryTimeMs = null;
         }
 
-        private boolean isResetAllowed(long nowMs) {
+        private synchronized boolean isResetAllowed(long nowMs) {
             return nextAllowedRetryTimeMs == null || nowMs >= nextAllowedRetryTimeMs;
         }
 
-        private boolean awaitingReset() {
+        private synchronized boolean awaitingReset() {
             return resetStrategy != null;
         }
 
-        private void setResetPending(long nextAllowedRetryTimeMs) {
+        private synchronized void setResetPending(long nextAllowedRetryTimeMs) {
             this.nextAllowedRetryTimeMs = nextAllowedRetryTimeMs;
         }
 
-        private void resetFailed(long nextAllowedRetryTimeMs) {
+        private synchronized void resetFailed(long nextAllowedRetryTimeMs) {
             this.nextAllowedRetryTimeMs = nextAllowedRetryTimeMs;
         }
 
-        private boolean hasValidPosition() {
+        private synchronized boolean hasValidPosition() {
             return position != null;
         }
 
-        private boolean isMissingPosition() {
+        private synchronized boolean isMissingPosition() {
             return !hasValidPosition() && !awaitingReset();
         }
 
-        private boolean isPaused() {
+        private synchronized boolean isPaused() {
```

Comment:
Do we still need synchronization here if SubscriptionState is synchronized?

Reply:
Not all method in SubscriptionState is synchronized. Basically when I don't need to modify `assignment` I only synchronize until I get the `TopicPartitionState` and then lock that object to avoid locking the whole collection when it's not necessary.
If you look at `isPaused` there:
```
public boolean isPaused(TopicPartition tp) {
        TopicPartitionState assignedOrNull = assignedStateOrNull(tp);
        return assignedOrNull != null && assignedOrNull.isPaused();
    }
```
the first line would only get the state from the collection, release the lock of `SubscriptionState` and only lock the given `TopicPartitionState`.
When I write the collection then I lock for instance in `assignFromUser` then I lock the whole method as I think we should keep writes consistent.
If you think this is unnecessary or complicated I can just lock `SubscriptionState` every time.

Reply:
Yeah, I have a slight preference to just lock `SubscriptionState` every time since it is the simplest option. I don't think contention is a major problem since there's only the heartbeat thread which is sleeping most of the time. Unless there's some reason to think the cost of lock acquisition itself is a concern.

Reply:
Ok. I don't think lock acquisition would be costy in this case.